File: Leetcode/138/138.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    unordered_map<Node*,Node*> node2node;
    Node* copyRandomList(Node* head) {
        if (!head) {
            return nullptr;
        }
        for (auto &[k,v] : node2node) {
            if (v) {
                delete v;
            }
        }
        node2node.clear();
        auto node(head);
        while (node) {
            node2node[node] = new Node(node->val);
            node = node->next;
        }
        node = head;
        while (node) {
            if (node->random) {
                node2node[node]->random = node2node[node->random];
            }
            node = node->next;
        }
        for (auto &[k,v] : node2node) {
            if (k->next) {
                v->next = node2node[k->next];
            }
        }
        return node2node[head];
    }
};

File: Leetcode/138/138.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}
	old2new := make(map[*Node]*Node)
	node := head
	beforeHead := new(Node)
	var currentNewNode *Node = nil
	for node != nil {
		temp := new(Node)
		temp.Val = node.Val
		temp.Next = nil
		temp.Random = nil
		if currentNewNode == nil {
			currentNewNode = temp
			beforeHead.Next = currentNewNode
		} else {
			currentNewNode.Next = temp
			currentNewNode = currentNewNode.Next
		}
		node = node.Next
	}
	node = head
	node2 := beforeHead.Next
	for node != nil {
		old2new[node] = node2
		node = node.Next
		node2 = node2.Next
	}
	node = head
	for node != nil {
		if node.Random != nil {
			old2new[node].Random = old2new[node.Random]	
		}
		node = node.Next
	}
	return beforeHead.Next
}